### PR TITLE
[Merged by Bors] - feat(CategoryTheory): the typeclass Functor.HasRightKanExtension

### PR DESCRIPTION
**Pull request by joelriou**
This PR introduces the typeclass `HasRightKanExtension L F` which asserts the existence of a right Kan extension of a functor `F` along `L`. A chosen extension is obtained as `rightKanExtension L F`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Equivalence.lean [hunk 1]
```diff
@@ -567,6 +567,14 @@ instance isEquivalenceTrans (F : C ⥤ D) (G : D ⥤ E) [IsEquivalence F] [IsEqu
   IsEquivalence.ofEquivalence (Equivalence.trans (asEquivalence F) (asEquivalence G))
 #align category_theory.functor.is_equivalence_trans CategoryTheory.Functor.isEquivalenceTrans
 
+instance (F : C ⥤ D) [IsEquivalence F] :
+    IsEquivalence ((whiskeringLeft C D E).obj F) :=
```
Comment (TwoFX) [inline Mathlib/CategoryTheory/Equivalence.lean:571]:
```suggestion
instance (F : C ⥤ D) [IsEquivalence F] : IsEquivalence ((whiskeringLeft C D E).obj F) :=
```

File: Mathlib/CategoryTheory/Equivalence.lean [hunk 2]
```diff
@@ -567,6 +567,14 @@ instance isEquivalenceTrans (F : C ⥤ D) (G : D ⥤ E) [IsEquivalence F] [IsEqu
...
+  (inferInstance : IsEquivalence (Equivalence.congrLeft F.asEquivalence).inverse)
+
+instance (F : C ⥤ D) [IsEquivalence F] :
+    IsEquivalence ((whiskeringRight E C D).obj F) :=
```
Comment (TwoFX) [inline Mathlib/CategoryTheory/Equivalence.lean:575]:
```suggestion
instance (F : C ⥤ D) [IsEquivalence F] : IsEquivalence ((whiskeringRight E C D).obj F) :=
```

File: Mathlib/CategoryTheory/Functor/KanExtension/Basic.lean [hunk 3]
```diff
@@ -169,6 +173,143 @@ lemma isLeftKanExtension_iff_of_iso {F' : H ⥤ D} {F'' : H ⥤ D} (e : F' ≅ F
...
+lemma hasLeftExtension_iff_postcomp₁ :
+    HasLeftKanExtension L F ↔ HasLeftKanExtension (L ⋙ e.functor) F :=
+  (leftExtensionEquivalenceOfPostcomp₁ L F e).symm.hasInitial_iff
+
+variable {L L'}
+
+/-- The equivalence `RightExtension L F ≌ RightExtension L' F` induced by
+a natural isomorphism `L ≅ L'`. -/
+def rightExtensionEquivalenceOfIso₁ :
+    RightExtension L F ≌ RightExtension L' F :=
```
Comment (TwoFX) [inline Mathlib/CategoryTheory/Functor/KanExtension/Basic.lean:272]:
```suggestion
def rightExtensionEquivalenceOfIso₁ : RightExtension L F ≌ RightExtension L' F :=
```

File: Mathlib/CategoryTheory/Functor/KanExtension/Basic.lean [hunk 4]
```diff
@@ -169,6 +173,143 @@ lemma isLeftKanExtension_iff_of_iso {F' : H ⥤ D} {F'' : H ⥤ D} (e : F' ≅ F
...
+  CostructuredArrow.mapNatIso ((whiskeringLeft C H D).mapIso iso₁)
+
+lemma hasRightExtension_iff_of_iso₁ :
+    HasRightKanExtension L F ↔ HasRightKanExtension L' F :=
```
Comment (TwoFX) [inline Mathlib/CategoryTheory/Functor/KanExtension/Basic.lean:276]:
```suggestion
lemma hasRightExtension_iff_of_iso₁ : HasRightKanExtension L F ↔ HasRightKanExtension L' F :=
```

File: Mathlib/CategoryTheory/Functor/KanExtension/Basic.lean [hunk 5]
```diff
@@ -169,6 +173,143 @@ lemma isLeftKanExtension_iff_of_iso {F' : H ⥤ D} {F'' : H ⥤ D} (e : F' ≅ F
...
+  (rightExtensionEquivalenceOfIso₁ iso₁ F).hasTerminal_iff
+
+/-- The equivalence `LeftExtension L F ≌ LeftExtension L' F` induced by
+a natural isomorphism `L ≅ L'`. -/
+def leftExtensionEquivalenceOfIso₁ :
+    LeftExtension L F ≌ LeftExtension L' F :=
```
Comment (TwoFX) [inline Mathlib/CategoryTheory/Functor/KanExtension/Basic.lean:282]:
```suggestion
def leftExtensionEquivalenceOfIso₁ : LeftExtension L F ≌ LeftExtension L' F :=
```

File: Mathlib/CategoryTheory/Functor/KanExtension/Basic.lean [hunk 6]
```diff
@@ -169,6 +173,143 @@ lemma isLeftKanExtension_iff_of_iso {F' : H ⥤ D} {F'' : H ⥤ D} (e : F' ≅ F
...
+  StructuredArrow.mapNatIso ((whiskeringLeft C H D).mapIso iso₁)
+
+lemma hasLeftExtension_iff_of_iso₁ :
+    HasLeftKanExtension L F ↔ HasLeftKanExtension L' F :=
```
Comment (TwoFX) [inline Mathlib/CategoryTheory/Functor/KanExtension/Basic.lean:286]:
```suggestion
lemma hasLeftExtension_iff_of_iso₁ : HasLeftKanExtension L F ↔ HasLeftKanExtension L' F :=
```

File: Mathlib/CategoryTheory/Functor/KanExtension/Basic.lean [hunk 7]
```diff
@@ -169,6 +173,143 @@ lemma isLeftKanExtension_iff_of_iso {F' : H ⥤ D} {F'' : H ⥤ D} (e : F' ≅ F
...
+  (leftExtensionEquivalenceOfIso₁ iso₁ F).hasInitial_iff
+
+variable (L) {F F'}
+
+/-- The equivalence `RightExtension L F ≌ RightExtension L F'` induced by
+a natural isomorphism `F ≅ F'`. -/
+def rightExtensionEquivalenceOfIso₂ :
+    RightExtension L F ≌ RightExtension L F' :=
```
Comment (TwoFX) [inline Mathlib/CategoryTheory/Functor/KanExtension/Basic.lean:294]:
```suggestion
def rightExtensionEquivalenceOfIso₂ : RightExtension L F ≌ RightExtension L F' :=
```

File: Mathlib/CategoryTheory/Functor/KanExtension/Basic.lean [hunk 8]
```diff
@@ -169,6 +173,143 @@ lemma isLeftKanExtension_iff_of_iso {F' : H ⥤ D} {F'' : H ⥤ D} (e : F' ≅ F
...
+  CostructuredArrow.mapIso iso₂
+
+lemma hasRightExtension_iff_of_iso₂ :
+    HasRightKanExtension L F ↔ HasRightKanExtension L F' :=
```
Comment (TwoFX) [inline Mathlib/CategoryTheory/Functor/KanExtension/Basic.lean:298]:
```suggestion
lemma hasRightExtension_iff_of_iso₂ : HasRightKanExtension L F ↔ HasRightKanExtension L F' :=
```

File: Mathlib/CategoryTheory/Functor/KanExtension/Basic.lean [hunk 9]
```diff
@@ -169,6 +173,143 @@ lemma isLeftKanExtension_iff_of_iso {F' : H ⥤ D} {F'' : H ⥤ D} (e : F' ≅ F
...
+  (rightExtensionEquivalenceOfIso₂ L iso₂).hasTerminal_iff
+
+/-- The equivalence `LeftExtension L F ≌ LeftExtension L F'` induced by
+a natural isomorphism `F ≅ F'`. -/
+def leftExtensionEquivalenceOfIso₂ :
+    LeftExtension L F ≌ LeftExtension L F' :=
```
Comment (TwoFX) [inline Mathlib/CategoryTheory/Functor/KanExtension/Basic.lean:304]:
```suggestion
def leftExtensionEquivalenceOfIso₂ : LeftExtension L F ≌ LeftExtension L F' :=
```

File: Mathlib/CategoryTheory/Functor/KanExtension/Basic.lean [hunk 10]
```diff
@@ -169,6 +173,143 @@ lemma isLeftKanExtension_iff_of_iso {F' : H ⥤ D} {F'' : H ⥤ D} (e : F' ≅ F
...
+  StructuredArrow.mapIso iso₂
+
+lemma hasLeftExtension_iff_of_iso₂ :
+    HasLeftKanExtension L F ↔ HasLeftKanExtension L F' :=
```
Comment (TwoFX) [inline Mathlib/CategoryTheory/Functor/KanExtension/Basic.lean:308]:
```suggestion
lemma hasLeftExtension_iff_of_iso₂ : HasLeftKanExtension L F ↔ HasLeftKanExtension L F' :=
```

**Review by TwoFX — APPROVED** [review]
Thanks!

The whitespace suggestions are all optional, whatever you prefer.

bors d+

**Comment by joelriou** [comment]
Thanks for the review!

bors merge

**Pull request closed without merging**